extensions/smallrye-fault-tolerance/runtime/src/main/java/io/quarkus/smallrye/faulttolerance/runtime/QuarkusFaultToleranceOperationProvider.java

```diff
@@ -1,22 +1,76 @@
 package io.quarkus.smallrye.faulttolerance.runtime;
 
 import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
 
 import javax.annotation.Priority;
-import javax.enterprise.context.Dependent;
 import javax.enterprise.inject.Alternative;
+import javax.inject.Singleton;
+
+import org.jboss.logging.Logger;
 
 import io.smallrye.faulttolerance.FaultToleranceOperationProvider;
 import io.smallrye.faulttolerance.config.FaultToleranceOperation;
 
-@Dependent
+@Singleton
 @Alternative
 @Priority(1)
 public class QuarkusFaultToleranceOperationProvider implements FaultToleranceOperationProvider {
+    private static final Logger LOG = Logger.getLogger(QuarkusFaultToleranceOperationProvider.class);
+
+    private final Map<CacheKey, FaultToleranceOperation> operationCache = new ConcurrentHashMap<>();
+    private final Function<CacheKey, FaultToleranceOperation> cacheFunction = new Function<CacheKey, FaultToleranceOperation>() {
+        @Override
+        public FaultToleranceOperation apply(CacheKey key) {
+            return createAtRuntime(key);
+        }
+    };
+
+    /**
+     * Called by SmallryeFaultToleranceRecorder to init the operation cache.
+     */
+    void init(Map<CacheKey, FaultToleranceOperation> operationCache) {
+        this.operationCache.putAll(operationCache);
+    }
 
     @Override
     public FaultToleranceOperation get(Class<?> beanClass, Method method) {
-        return FaultToleranceOperation.of(beanClass, method);
+        CacheKey key = new CacheKey(beanClass, method);
+        FaultToleranceOperation existing = operationCache.get(key);
+        return existing != null ? existing : operationCache.computeIfAbsent(key, cacheFunction);
+    }
+
+    private FaultToleranceOperation createAtRuntime(CacheKey key) {
+        LOG.debugf("FaultToleranceOperation not found in the cache for %s creating it at runtime", key);
+        return FaultToleranceOperation.of(key.beanClass, key.method);
     }
 
+    static class CacheKey {
+        private Class<?> beanClass;
+        private Method method;
+
+        public CacheKey(Class<?> beanClass, Method method) {
+            this.beanClass = beanClass;
+            this.method = method;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o)
+                return true;
+            if (o == null || getClass() != o.getClass())
+                return false;
```

Comment:
If you end up taking into account my comment below, please use curly braces for the two statements above.

Reply:
Done, I should update my IDEA templates as I prefere having the curly braces ...